tianshou/utils/net/discrete.py

```diff
@@ -111,3 +111,106 @@ def forward(
         """Mapping: s -> V(s)."""
         logits, _ = self.preprocess(s, state=kwargs.get("state", None))
         return self.last(logits)
+
+
+class CosineEmbeddingNetwork(nn.Module):
+    """Cosine embedding network for IQN. Convert a scalar in [0, 1] to a list
+    of n-dim vectors.
+
+    :param num_cosines: the number of cosines used for the embedding.
+    :param embedding_dim: the dimension of the embedding/output.
+
+    .. note::
+
+        From https://github.com/ku2482/fqf-iqn-qrdqn.pytorch/blob/master
+        /fqf_iqn_qrdqn/network.py .
+    """
+    def __init__(self, num_cosines: int, embedding_dim: int) -> None:
+        super().__init__()
+        self.net = nn.Sequential(nn.Linear(num_cosines, embedding_dim), nn.ReLU())
+        self.num_cosines = num_cosines
+        self.embedding_dim = embedding_dim
+
+    def forward(self, taus: torch.Tensor) -> torch.Tensor:
+        batch_size = taus.shape[0]
+        N = taus.shape[1]
+        # Calculate i * \pi (i=1,...,N).
+        i_pi = np.pi * torch.arange(
+            start=1, end=self.num_cosines + 1, dtype=taus.dtype, device=taus.device
+        ).view(1, 1, self.num_cosines)
+        # Calculate cos(i * \pi * \tau).
+        cosines = torch.cos(taus.view(batch_size, N, 1) * i_pi).view(
+            batch_size * N, self.num_cosines
+        )
+        # Calculate embeddings of taus.
+        tau_embeddings = self.net(cosines).view(batch_size, N, self.embedding_dim)
+        return tau_embeddings
+
+
+class IQN(Critic):
```

Comment:
I personally recommend moving this network under test_iqn.py because few algorithms reuse this network architecture. But anyway it's just my personal thought, it's okay if you insist.

Reply:
I agree that it may not be "common" enough to put it here but I feel putting it in `test_iqn.py` is not appropriate, either, because I need it in `atari_iqn.py` and importing from a test file is very weird. The other way around (`test_iqn.py` importing this class from `atari_iqn.py`) is also awkward. Therefore, before we find a more suitable place for this class, I'd rather let it be here.

I am not sure about the naming. Do you think it'd be better if I change it from `IQN` to `ImplicitQuantileNetwork`?